kubeadm: Perform IPVS check on join only if the cluster was set up with IPVS

ereslibre opened the pull request:
This change will also allow any join operation to discover the init configuration
of the cluster before running the join preflight checks, so these join preflight
checks can depend on what configuration kubeadm had when deployed.

**What this PR does / why we need it**:
This PR allows the join process to download the `InitConfiguration` from the cluster based on the `JoinConfiguration`, discovering the `InitConfiguration` before we set up the preflight checks, so those checks can honor this configuration and perform extra checks or ignore others depending on the `InitConfiguration` of the cluster.

**Which issue(s) this PR fixes**:
Fixes https://github.com/kubernetes/kubeadm/issues/1128

**Special notes for your reviewer**:
None

**Release note**:
```release-note
NONE
```

ereslibre committed:
kubeadm: Perform IPVS check on join only if the cluster was set up with IPVS.

neolit123 commented:
/cc

fabriziopandini commented:
@ereslibre I will jump on this on Monday
/assign @fabriziopandini

ereslibre commented:
Thank you @fabriziopandini. I'm currently with a very bad connection on holidays and having a hard time verifying things. I'll also get back to it on Monday.

fabriziopandini reviewed (changes requested):
@ereslibre 
I appreaciate the fact that you are sharing this WIP! 
I think that the responsibility of reading initconfiguration should remain in charge of join.

So I kindly ask you to try a slightly different approach where you read initconfiguration in NewJoin (before calling preflight).

ereslibre commented:
@fabriziopandini Thank you for the instructions. The only issue I see on reading the `InitConfiguration` on the `NewJoin` is that we also have https://github.com/kubernetes/kubernetes/blob/02cca1f11a8ae570369540564604d0cde06e0720/cmd/kubeadm/app/cmd/phases/preflight.go#L138; this `alpha phase preflight node` would also need to download the `InitConfiguration` in order to call to `RunJoinNodeChecks` with the `InitConfiguration`.

If that's fine I can see how to implement it this way without code repetition and also avoiding import cycles.

neolit123 commented:
might be a good idea to make "init from join" into a utility instead of keeping it in join.go or checks.go.

it's also needed in one other case:
cmd/kubeadm/app/cmd/phases/kubelet.go

https://github.com/kubernetes/kubernetes/pull/66798/files#diff-4e06863a72cb928f527a6d3a8ad4ec5bR164

ereslibre commented:
> might be a good idea to make "init from join" into a utility instead of keeping it in join.go or checks.go.

Yes, I'm going to explore this possibility.

ereslibre commented:
Patch updated; please note that once that we're happy with the results I'll squash the commits.

I'm going to double check that it behaves as expected; it could also use some unit testing.

neolit123 commented:
SGTM, i like having this as a utility.
@fabriziopandini WDYT?

ereslibre commented:
Hmm, just found out that it's trying to download the `kubelet-config-1.11` when in reality `kubelet-config-1.12` exists:

```
kubic-seed:~ # kubectl get configmap --all-namespaces
NAMESPACE     NAME                                 DATA      AGE
kube-public   cluster-info                         2         7m41s
kube-system   coredns                              1         7m41s
kube-system   extension-apiserver-authentication   6         7m50s
kube-system   kube-proxy                           2         7m41s
kube-system   kubeadm-config                       2         7m43s
kube-system   kubelet-config-1.12                  1         7m43s
```

Hmm, I wonder why `kubeadm` tries to download the `kubelet-config-1.11` instead of `kubelet-config-1.12`. I used the same `kubeadm` version to bootstrap the first node:

```
./kubeadm init --cri-socket=/var/run/crio/crio.sock --pod-network-cidr=10.244.0.0/16
```

and then on the new machine:

```
./kubeadm join 192.168.122.68:6443 --token qrw4jk.tj9osoumt850pkr9 --discovery-token-ca-cert-hash sha256:390f161bf119fff984dec6eba37d296be6bd47f632d411edca8c4d37b8f1e3bc --cri-socket=/var/run/crio/crio.sock
```

and been hitting this error. I'll keep digging.

ereslibre commented:
Nevermind, `preflight.GetKubeletVersion` will parse `kubelet --version` output, and I wasn't uploading the new `kubelet` executable to the machines but using an old one.

fabriziopandini commented on cmd/kubeadm/app/util/config/initconfiguration.go:
```diff
@@ -21,26 +21,72 @@ import (
+	// Perform the Discovery, which turns a Bootstrap Token and optionally (and preferably) a CA cert hash into a KubeConfig
+	// file that may be used for the TLS Bootstrapping process the kubelet performs using the Certificates API.
+	glog.V(1).Infoln("[join] discovering cluster-info")
+	tlsBootstrapCfg, err := discovery.For(cfg)
```
I don't think this function belongs here.

`discovery` is

fabriziopandini reviewed (changes requested):
@ereslibre 
I'm not really concerned about the `preflight node phase`, because ATM this is the only existing phase for the join workflow and this doesn't really make sense. IMO, such phase should a) go away in the context of the phase graduation activity we are executing or b) be refactored using the new phaseRunner recently merged, but this is not yet decided.

So, for the time being, I will be happy to keep the discovery/fetch code in join.go; the `preflight node phase`, with this approach doesn't get the initcfg/kube proxy, can either continue to always execute the check as of today or change the behavior and never run it

fabriziopandini commented:
/ok-to-test

neolit123 reviewed (commented):
if there are concerns about this PR, moving the FetchInitConfiguration*() should go in a separate PR at least. i think we need this in more than one places.

ereslibre commented:
@fabriziopandini as per @neolit123's comment, is it OK if we leave `FetchInitConfiguration{,FromJoinConfiguration}` where it is at the moment, and adapt `preflight node phase` to not pass any `initCfg`, making this test always execute? If other parts of `kubeadm` are going to need these two functions I think it makes sense to leave them as utilities?

If after other refactors they don't make sense as a util, and join is the only user I think we can refactor this later and move it inside `join` afterwards?

Please confirm what solution do you prefer, so I can fix the tests and make the PR green.

ereslibre commented:
> i guess my point was that we want (at least i do) the Fetch* functions to move

Yeah, with `where it is at the moment` I meant `where it is at the moment in this PR`, sorry for the confusion. The rest of my message goes in this direction too: I'm suggesting to keep it as a util for now, until further refactors render `join` as its only user and then we could refactor them and move them out of utils if the time comes.

ereslibre commented:
Added a new commit to pass `nil` when we aren't interested in fetching the init config and to fix tests.

fabriziopandini commented:
@ereslibre @neolit123 
I'm still of the opinion that a piece of code the includes call to discovery, usage of a TLSbootstrap and merge between Init Configuration and Join Configuration is very specific of the join workflow and it should be in join.go

Refactoring the code above in something more generic that can be used elsewhere/in future IMO is out of scope of this PR

neolit123 commented:
@fabriziopandini 
> I'm still of the opinion that a piece of code the includes call to discovery, usage of a TLSbootstrap and merge between Init Configuration and Join Configuration is very specific of the join workflow and it should 
be in join.go

if there are more consumers of such a function would they have to import cmd/join.go?

could you please look at this PR and provide assessment:
https://github.com/kubernetes/kubernetes/pull/66798/files#diff-4e06863a72cb928f527a6d3a8ad4ec5bR164

basically i need to fetch the ClusterConfiguration from JoinConfiguration in cmd/kubeadm/app/cmd/phases/kubelet.go as well. this was discussed with @rosti last week.

fabriziopandini commented:
@neolit123 
In the context of this PR I don't see this change fits and I explained why.
If another PR needs to create a utility function, IMO it isthis second PR that should take care of this so the reviewer will get all the context to give feedbacks

ereslibre commented:
@neolit123 @fabriziopandini I will move the logic to `join.go`; other changes can refactor this if they need to.

neolit123 commented:
@ereslibre @fabriziopandini 
ok, let's get this merged.

/retest

ereslibre commented:
@fabriziopandini @neolit123 Squashed and pushed all the changes. Now this utility functions live inside `app/cmd/util/join.go`. I hope this location is more appropriate, at least for the time being.

ereslibre commented:
/test pull-kubernetes-verify

neolit123 commented:
@ereslibre
we are going back and forth here. :-)

> Now this utility functions live inside app/cmd/util/join.go. I hope this location is more appropriate, at least for the time being.

@fabriziopandini requested that we should not move the utility in this PR.

> Refactoring the code above in something more generic that can be used elsewhere/in future IMO is out of scope of this PR

ereslibre commented:
@neolit123 In this PR, these two functions are in `cmd/kubeadm/app/cmd/util/join.go`. Should I move these two functions to `cmd/kubeadm/app/cmd/join.go`? Note that in my first attempt, that @fabriziopandini requested changes upon, I added it to `cmd/kubeadm/app/util/config/initconfiguration.go`.

That's why I'm a bit confused; @fabriziopandini referred to adding these two functions inside `join.go`, and I thought `app/cmd/util/join.go` was the place he was referring to. If you prefer it to go inside `app/cmd/join.go`, please confirm and I'll update it, no problem :)

neolit123 commented:
well, i think @fabriziopandini meant to not move the function from cmd/join.go
let's wait on his opinion again.

ereslibre commented:
@neolit123 Rereading I think you're right; sorry for the mess. PR updated.

timothysc commented:
/assign @timothysc

neolit123 commented on cmd/kubeadm/app/cmd/join.go:
```diff
@@ -260,6 +260,7 @@ func AddJoinOtherFlags(flagSet *flag.FlagSet, cfgPath *string, ignorePreflightEr
 // Join defines struct used by kubeadm join command
 type Join struct {
 	cfg                   *kubeadmapi.JoinConfiguration
+	initCfg               *kubeadmapi.InitConfiguration
```
having both configs in a struct called Join feels weird indeed. the struct is just used for storage in the case of initcfg, so that fetchInitConfigurationFromJoinConfiguration() is not called multiple times.
if they remain like so they should be joinCfg, initCfg. also line 299 would need some renames.

on control-plane nodes the IPVS check is conditional on kube-proxy mode already. the conditional IPVS check on worker nodes seems viable in cases of worker nodes not having the IPVS modules where an error will always be thrown.

fabriziopandini replied:
@timothysc what we really need here is ClusterConfiguration but as of today InitConfiguration and ClusterConfiguration are still tied together in many utilities.
Fixing the above point IMO is out scope of this PR, so I'm fine with the proposed solution with one last exception already discussed with @ereslibre

neolit123 commented on cmd/kubeadm/app/cmd/join.go:
```diff
@@ -282,15 +283,20 @@ func NewJoin(cfgPath string, args []string, defaultcfg *kubeadmapiv1beta1.JoinCo
 		return nil, err
 	}
 
+	initcfg, err := fetchInitConfigurationFromJoinConfiguration(internalcfg)
```
one interesting question here (to me at least), with this we are trying to fetch the init configuration, **before** we have checked if this node is healthy.
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/preflight/checks.go#L933

is this a good idea?

ereslibre replied:
Another option would be to split the checks into two groups, we run the general join node checks; if everything is fine we then fetch the init configuration and run the checks that apply to that init configuration, kind of splitting `RunJoinNodeChecks` into `RunPreJoinNodeChecks` and `RunPostJoinNodeChecks` or something similar?

fabriziopandini commented on cmd/kubeadm/app/cmd/join.go:
```diff
@@ -282,15 +283,20 @@ func NewJoin(cfgPath string, args []string, defaultcfg *kubeadmapiv1beta1.JoinCo
 		return nil, err
 	}
 
+	initcfg, err := fetchInitConfigurationFromJoinConfiguration(internalcfg)
```
@ereslibre please fix the fact we are doing TLSbootstrap twice

ereslibre replied:
Yes, I will have a look at this today. Thanks!

ereslibre commented:
@fabriziopandini @neolit123 I refactored this PR a bit; checks are done into two phases: general unconditional checks, fetch the init configuration and then conditional checks based on this configuration that can grow over time. I also removed the several discoveries that I was doing here.

neolit123 commented on cmd/kubeadm/app/cmd/join.go:
```diff
@@ -303,52 +305,42 @@ func NewJoin(cfgPath string, defaultcfg *kubeadmapiv1beta1.JoinConfiguration, ig
-		if err != nil {
-			return err
-		}
+	glog.V(1).Infoln("[preflight] running specific configuration-dependant checks")
```
omit the word `specific`?

neolit123 commented on cmd/kubeadm/app/cmd/join.go:
```diff
@@ -624,3 +599,44 @@ func waitForTLSBootstrappedClient() error {
+	// creates a client to access the cluster using the bootstrap token identity
+	tlsClient, err := kubeconfigutil.ToClientSet(tlsBootstrapCfg)
+	if err != nil {
+		return nil, errors.Wrap(err, "Unable to access the cluster")
```
error should start with lowercase.

neolit123 commented on cmd/kubeadm/app/cmd/join.go:
```diff
@@ -624,3 +599,44 @@ func waitForTLSBootstrappedClient() error {
+	// Fetches the init configuration
+	initConfiguration, err := configutil.FetchConfigFromFileOrCluster(tlsClient, os.Stdout, "join", "", true)
+	if err != nil {
+		return nil, errors.Wrap(err, "Unable to fetch the kubeadm-config ConfigMap")
```
error should start with lowercase.

neolit123 reviewed (commented):
@ereslibre thank you.
added some minor comments.

neolit123 commented:
it's a similar case to this:
https://github.com/kubernetes/kubeadm/issues/1189

"kubeadm join for worker should not require identifying a bind address"

fabriziopandini commented:
@neolit123 I don't think 1189 is related to this issue (that logic was there before and this PR is not changing it)

@timothysc I will contact you directly to see if we can find a way forward

neolit123 commented:
@fabriziopandini i meant both issues are about worker nodes that should be doing conditional checks.

fabriziopandini commented:
@ereslibre 
Thanks for this PR and all your effort in addressing all the comments!
Other issue are will take charge of further code cleanups in the join workflow and/or for the fetch configuration part (rif issues https://github.com/kubernetes/kubeadm/issues/1204 or https://github.com/kubernetes/kubeadm/issues/1209)

/approve
/lgtm
